test/extensions/filters/http/ext_proc/filter_test.cc

```diff
@@ -2261,6 +2272,198 @@ TEST_F(HttpFilterTest, FailOnInvalidHeaderMutations) {
+
+        // Add ext_proc filter.
+        Http::FilterFactoryCb cb = [&](Http::FilterChainFactoryCallbacks& callbacks) {
+          callbacks.addStreamFilter(filter_);
+        };
+        manager.applyFilterFactoryCb({}, cb);
+        // Add the mock-decoder filter.
+        Http::FilterFactoryCb mock_decode_filter_cb =
+            [&](Http::FilterChainFactoryCallbacks& callbacks) {
+              callbacks.addStreamDecoderFilter(mock_decode_filter);
+            };
+        manager.applyFilterFactoryCb({}, mock_decode_filter_cb);
+
+        return true;
+      }));
+  EXPECT_CALL(*mock_decode_filter, decodeHeaders(_, _))
+      .WillOnce(Invoke([&](Http::RequestHeaderMap& headers, bool end_stream) {
+        EXPECT_TRUE(end_stream);
+        EXPECT_EQ(headers.Path()->value().getStringView(), "/bluh");
+        // Direct response from decode filter.
+        Http::ResponseHeaderMapPtr response_headers{
+            new Http::TestResponseHeaderMapImpl{{":status", "200"}, {"foo", "foo-value"}}};
+        mock_decode_filter->callbacks_->encodeHeaders(std::move(response_headers), false,
+                                                      "filter_direct_response");
+        // Send a large body in one shot.
+        Buffer::OwnedImpl fake_response(
+            "Direct response from mock filter, Definitely more than 10 bytes data.");
+        mock_decode_filter->callbacks_->encodeData(fake_response, true);
+
+        // Now return from ext server the response for processing response headers.
+        EXPECT_TRUE(last_request_.has_response_headers());
+        auto response = std::make_unique<ProcessingResponse>();
+        auto* headers_response = response->mutable_response_headers();
+        auto* hdr =
+            headers_response->mutable_response()->mutable_header_mutation()->add_set_headers();
+        hdr->mutable_header()->set_key("foo");
+        hdr->mutable_header()->set_value("gift-from-external-server");
+        hdr = headers_response->mutable_response()->mutable_header_mutation()->add_set_headers();
+        hdr->mutable_header()->set_key("new_response_header");
+        hdr->mutable_header()->set_value("bluh");
+        HttpFilterTest::stream_callbacks_->onReceiveMessage(std::move(response));
+        return Http::FilterHeadersStatus::StopIteration;
+      }));
+  // Start the request.
+  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance& data) -> Http::Status {
+    EXPECT_EQ(data.length(), 5);
+    data.drain(5);
+    HttpConnectionManagerImplMixin::decoder_ = &conn_manager_->newStream(response_encoder_);
+    Http::RequestHeaderMapPtr headers{new Http::TestRequestHeaderMapImpl{
+        {":authority", "host"}, {":path", "/bluh"}, {":method", "GET"}}};
+    HttpConnectionManagerImplMixin::decoder_->decodeHeaders(std::move(headers), true);
+    return Http::okStatus();
+  }));
+
+  Buffer::OwnedImpl fake_input("hello");
+  conn_manager_->onData(fake_input, false);
```

Comment:
Similar to the comment for `LastDecodeDataCallExceedsStreamBufferLimitWouldJustRaiseHighWatermark`: what part of this test "proves" the correct behavior? The test comment says we avoid a 500 error, but AFAICT the test does not check the client-visible response.

Maybe the "proof" is that some high-level watermark callback isn't invoked? If so, call that out in a comment, since it's not obvious to me where it is verified in the test.

Reply:
see the comment added around line: 2387.
I also checked the status there.

Reply:
LG